File: playground/mechanics_qa.py
"""Versioned physical experiments; user-edited trials never replace QA fixtures."""
from copy import deepcopy
from pathlib import Path
import threading
import time
import uuid
import material_qa as artifacts
import physics_trials

ROOT = Path(__file__).resolve().parents[1]
FIXTURES = ROOT / "examples/physics-trials/regression.json"
_LOCK = threading.Lock()

def cases():
    return artifacts.read_json(FIXTURES)["cases"]

def select(ids=None):
    known = {c["id"]: c for c in cases()}
    if ids is None: return list(known.values())
    if not isinstance(ids, list) or not ids or len(ids) > len(known) or any(not isinstance(i, str) or i not in known for i in ids) or len(set(ids)) != len(ids):
        raise ValueError("case_ids must be a unique nonempty list from the mechanics catalog")
    return [known[i] for i in ids]

def binary(engine):
    if engine is None: return None
    p = Path(engine)
    return p.with_name("banjo_live_world_run" + (".exe" if p.suffix == ".exe" else ""))

def catalog(engine):
    return {**physics_trials.catalog(), "cases": cases(), "suite_hash": artifacts.digest(cases()),
            "engine_available": bool(binary(engine) and binary(engine).is_file()),
            "acceptance": "Versioned physical invariants, not automatically learned outcomes. Custom checks are authored assertions, not qualification."}

def run_suite(engine, directory, chosen=None, *, document=None, cancel=None):
    selected = [{"id": "custom", "group": "Custom experiment", "document": document}] if document is not None else select(chosen)
    selected = deepcopy(selected)
    for c in selected: c["document"], _ = physics_trials.validate(c["document"])
    directory = Path(directory).resolve()
    directory.mkdir(parents=True, exist_ok=False)
    report = {"schema": "banjo.mechanics-qa.v1", "id": directory.name, "status": "running",
              "kind": "custom" if document is not None else "regression", "total": len(selected),
              "completed": 0, "started_unix_s": time.time(), "results": [],
              "suite_hash": artifacts.digest(selected), "provenance": artifacts.provenance(binary(engine))}
    artifacts.write_json(directory/"report.json", report)
    for c in selected:
        if cancel and cancel.is_set(): break
        report["active_case"] = c["id"]
        artifacts.write_json(directory/"report.json", report)
        result = physics_trials.run(engine, directory/c["id"], c["document"], cancel=cancel)
        result.update(id=c["id"], group=c["group"])
        artifacts.write_json(directory/c["id"]/"result.json", result)
        report["results"].append(result); report["completed"] += 1
        artifacts.write_json(directory/"report.json", report)
    report.pop("active_case", None)
    report["status"] = ("cancelled" if cancel and cancel.is_set() else
                        "passed" if all(r["status"] == "passed" for r in report["results"]) else "failed")
    report["finished_unix_s"] = time.time()
    artifacts.write_json(directory/"report.json", report)
    return report

class Manager(artifacts.Manager):
    def __init__(self, app):
        super().__init__(app)
        self.root = Path(app.runs_path)/"mechanics-qa"

    def start(self, body):
        if not isinstance(body, dict) or set(body)-{"case_ids", "document"} or ("case_ids" in body and "document" in body):
            raise ValueError("Use case_ids or a document, not both")
        if "document" in body:
            document, _ = physics_trials.validate(body["document"])
            count = 1
        else:
            document = None; count = len(select(body.get("case_ids")))
        if not binary(self.app.engine_path) or not binary(self.app.engine_path).is_file(): raise ValueError("Build the native live world runner")
        with self.lock:
            if self.thread and self.thread.is_alive(): raise ValueError("A mechanics trial is already running")
            self.run_id = uuid.uuid4().hex; self.cancel_event = threading.Event()
            self.pending = {"id": self.run_id, "status": "starting", "total": count, "completed": 0,
                            "kind": "custom" if document is not None else "regression", "results": []}
            folder = self.folder(self.run_id)
            def work():
                try:
                    run_suite(self.app.engine_path, folder, body.get("case_ids"),
                              document=document, cancel=self.cancel_event)
                except Exception as exc:
                    artifacts.write_json(folder/"report.json", {**self.pending, "status": "failed", "error": str(exc)})
            self.thread = threading.Thread(target=work, name="banjo-mechanics-qa", daemon=True)
            self.thread.start()
            return dict(self.pending)

    def case(self, run_id, case_id, playback=False, request=False):
        if case_id not in {"custom"} | {c["id"] for c in cases()}: raise ValueError("Unknown mechanics case")
        filename = "request.json" if request else "playback.json" if playback else "result.json"
        return artifacts.read_json(self.folder(run_id)/case_id/filename)

def manager(app):
    with _LOCK:
        if not hasattr(app, "mechanics_qa"): app.mechanics_qa = Manager(app)
        return app.mechanics_qa

File: playground/mechanics-qa.js
import "/scene.js";
const $=id=>document.getElementById(id);
let token, catalog, runId="", report, viewer, serial=0, currentCase="", refreshing=null;
const active=()=>["starting","running"].includes(report?.status);
async function api(path,body){
 const r=await fetch(path,body===undefined?{}:{method:"POST",headers:{"Content-Type":"application/json","X-Banjo-Token":token},body:JSON.stringify(body)});
 const d=await r.json();if(!r.ok||d.error)throw Error(d.error||"Request failed");return d;
}
const tell=e=>{$("progress").textContent=e.message||String(e);};
function option(select,value,text){const o=document.createElement("option");o.value=value;o.textContent=text;select.append(o);}
function controls(){for(const id of ["run-all","run-custom"])$(id).disabled=active()||!catalog?.engine_available;$("cancel").disabled=!active();}
function clear(){++serial;currentCase="";viewer?.dispose();viewer=null;$("stage").replaceChildren();$("checks").replaceChildren();$("failures").textContent=$("evidence").textContent=$("outcome").textContent="";$("play").disabled=$("frame").disabled=true;}
async function list(preferred){
 const selected=runId;const data=await api("/api/mechanics-qa/runs");if(runId!==selected)return;$("runs").replaceChildren();
 for(const r of data.runs)option($("runs"),r.id,new Date((r.started_unix_s||Date.now()/1000)*1000).toLocaleString()+" · "+r.completed+"/"+r.total+" · "+r.status);
 runId=preferred||runId||data.runs.find(r=>r.total===catalog.cases.length)?.id||data.runs[0]?.id||"";$("runs").value=runId;
}
async function inspect(id){
 const request=++serial,run=runId;currentCase=id;
 viewer?.dispose();viewer=null;$("stage").replaceChildren();$("checks").replaceChildren();$("play").disabled=$("frame").disabled=true;
 $("outcome").textContent="Loading recorded result…";$("failures").textContent=$("evidence").textContent="";
 const r=await api("/api/mechanics-qa/runs/"+run+"/"+id);
 if(request!==serial||run!==runId)return;
 $("case-title").textContent=r.title;
 $("outcome").textContent=(report.kind==="custom"?"Authored checks: ":"Regression checks: ")+r.status;
 $("outcome").className=r.status;
 for(const c of r.checks||[]){const tr=document.createElement("tr");for(const text of [(c.body?c.body+" · ":"")+c.metric,Number(c.value).toPrecision(6),c.min+" … "+c.max,c.passed?"Pass":"FAIL"]){const td=document.createElement("td");td.textContent=text;tr.append(td);}tr.className=c.passed?"passed":"failed";$("checks").append(tr);}
 $("failures").textContent=r.error||((r.joints||[]).filter(j=>!j.attached).map(j=>j.a+" → "+j.b+": "+j.parted_because).join("\n"));
 $("evidence").textContent=JSON.stringify({measured:r.measured,limits:r.limitations,provenance:report.provenance,request_hash:r.request_hash},null,2);
 const playback=await api("/api/mechanics-qa/runs/"+run+"/"+id+"/playback");
 if(request!==serial||run!==runId||!playback.frames?.length)return;
 viewer=window.BanjoScene.create($("stage"),{frameAllPoses:true,onFrame:({index,frame})=>{$("frame").value=index;$("clock").textContent=Number(frame.time_s).toFixed(3)+" s";},onPlayState:playing=>{$("play").textContent=playing?"Pause":"Play";}});
 viewer.load(playback);viewer.setSpeed(Number($("speed").value));
 $("frame").max=playback.frames.length-1;$("play").disabled=$("frame").disabled=false;
}
async function refresh(){
 if(!runId||refreshing===runId)return;const id=runId;refreshing=id;
 try{const next=await api("/api/mechanics-qa/runs/"+id);if(id!==runId)return;const wasActive=active();report=next;
 if(wasActive&&!active())await list(id);if(id!==runId)return;
 $("progress").textContent=report.completed+"/"+report.total+" · "+report.status+(report.kind==="custom"?" · custom experiment":" · regression suite")+(report.error?" · "+report.error:"");
 $("case").replaceChildren();for(const c of report.results||[])option($("case"),c.id,c.title+" · "+c.status);
 if(currentCase)$("case").value=currentCase;
 controls();if(!currentCase&&report.results?.length)await inspect(report.results[0].id);
 }finally{if(refreshing===id)refreshing=null;}
}
async function start(body){const r=await api("/api/mechanics-qa/run",body);clear();report=r;runId=r.id;await list(runId);await refresh();controls();}
$("example").onchange=()=>{$("recipe").value=JSON.stringify(catalog.cases.find(c=>c.id===$("example").value).document,null,2);$("validation").textContent="";};
$("run-all").onclick=()=>start({}).catch(tell);
$("run-custom").onclick=()=>{try{start({document:JSON.parse($("recipe").value)}).catch(tell);}catch(e){tell(e);}};
$("validate").onclick=async()=>{try{await api("/api/mechanics-qa/validate",{document:JSON.parse($("recipe").value)});$("validation").textContent="Valid recipe. Run it to measure the physical result.";}catch(e){$("validation").textContent=e.message;}};
$("cancel").onclick=()=>api("/api/mechanics-qa/cancel",{run_id:runId}).then(()=>{$("cancel").disabled=true;}).catch(tell);
$("runs").onchange=()=>{clear();runId=$("runs").value;refresh().catch(tell);};
$("case").onchange=()=>inspect($("case").value).catch(tell);
$("play").onclick=()=>{if(viewer.frame===viewer.frameCount-1)viewer.reset();viewer.play();};
$("frame").oninput=()=>viewer?.setFrame(Number($("frame").value));
$("speed").onchange=()=>viewer?.setSpeed(Number($("speed").value));
try{token=(await api("/api/status")).csrf_token;catalog=await api("/api/mechanics-qa");
for(const c of catalog.cases)option($("example"),c.id,c.group+" · "+c.document.title);
$("example").onchange();await list();await refresh();controls();
if(!runId)$("progress").textContent="Choose a recipe or run the regression suite.";
setInterval(()=>{if(active()||report?.status==="unattached")refresh().catch(tell);},1000);
}catch(e){tell(e);}

$("ask").onsubmit=async event=>{
 event.preventDefault();const button=$("ask-send");button.disabled=true;
 $("validation").textContent="The lab assistant is editing the recipe…";
 try{const answer=await api("/api/mechanics-qa/plan",{message:$("ask-text").value,document:JSON.parse($("recipe").value)});
 if(answer.document)$("recipe").value=JSON.stringify(answer.document,null,2);
 $("validation").textContent=answer.explanation+(answer.document?" Recipe validated; run it to measure the result.":"");
 }catch(e){$("validation").textContent=e.message;}finally{button.disabled=false;}
};

try {
 const c=await api("/api/gameplay/capabilities");
 $("capability-count").textContent=c.counts.complete+"/30 complete · "+c.counts.partial+" partial · "+c.counts.planned+" planned";
 for(const item of c.items){const tr=document.createElement("tr");
 for(const text of [item.workstream,item.id+". "+item.name,item.status,item.remaining+(item.evidence?" Evidence: "+item.evidence:"")]){
 const td=document.createElement("td");td.textContent=text;tr.append(td);}
 $("capabilities").append(tr);}
}catch(e){$("capability-count").textContent=e.message;}
